### ISSUE TITLE
Training multi-task network with mostly missing labels

### ISSUE THREAD

**Issue by iskandr** [fchollet/keras#3206]
Sorry if this is something that's become more obvious with the functional API but I'm trying to train networks with >100 outputs and for each sample I only get labels for a random subset of those outputs (often more than 1 but rarely more than 10). Do I need to use a custom loss to zero out loss contributions from missing outputs or is somehow already handled within the `Model` class? 

Thanks!

**Comment by fchollet** [comment]
You could consider concatenating your outputs into a single output (e.g. if
you have 100 softmaxes, you would get a 3D output (batch_size, 100,
classes)) and use a masking layer + a softmax+categorical_crossentropy
loss.

softmax+categorical_crossentropy will work for 3D outputs (applied to the
last dimension), and your missing outputs (axis #1) would not contribute to
the loss if they are masked.

To pass a mask, you could consider having a custom masking layer connected
to a new model input that would feed with a mask tensor (1s and 0s). The
layer would return its inputs unchanged (the output of your network) but
would add a mask, which would be taken into account by the loss.

**Comment by sileod** [comment]
Isn't it better to make one model per output, and share the same parameters ?

**Comment by iskandr** [comment]
@fchollet Would that also work with an MSE loss for multiple regression targets?

Also, do you think something like this work for masking the output values? 

`Merge([network_outputs, Masking(mask_value=0)(mask_input)], mode=lambda xs: xs[0], output_mask=lambda xs: xs[1])`

**Comment by iskandr** [comment]
@sileod For samples with multiple labels I don't think I'd be able to correctly update the parameters when using multiple models. The desired behavior would be to get averaging of gradients from multiple outputs but with more than one model I would instead have to pick an order of sequentialized weight updates.

**Comment by fchollet** [comment]
> Isn't it better to make one model per output, and share the same
parameters ?

That's a good idea. It may be easier to manage.

> For samples with multiple labels I don't think I'd be able to correctly
update the parameters when using multiple models. The desired behavior
would be to get averaging of gradients from multiple outputs but with more
than one model I would instead have to pick an order of sequentialized
weight updates.

Picking a random order would work. It should not significantly hurt
training.

Also, your code looks good to me. That's the idea.

**Comment by iskandr** [comment]
Thanks for the feedback. I'll close this issue while I try both approaches and update later with how they worked out.

**Comment by giladdiv** [comment]
@iskandr  can you explain your mask function?

and also, I have a network with N classes and i want to evluate 3 parameters so I want to build 3 different softmax over the last layer which will be concatenation of fc layer with one-hot vector indicates the class. Does it sound reasonable? how can I build the 3 softmax outputs?

Thanks
